hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/OzoneManagerServiceProviderImpl.java

```diff
@@ -255,6 +265,22 @@ public void start() {
         TimeUnit.MILLISECONDS);
   }
 
+  private void stopSyncDataFromOMThread() {
+    scheduler.shutdownNow();
```

Comment:
Need to add another few lines of comments here:

1. `shutdownNow()` does not wait for executing task to terminate;
2. `stopSyncDataFromOMThread()` should only be called when no sync thread is actively running

Reply:
Let's add a sanity check here if possible. e.g. have a `Precondition` check here that the return value of shutdownNow() should be an empty list:

```suggestion
    Preconditions.checkState(scheduler.shutdownNow().isEmpty(),
        "OM DB sync is still running!");
```

Reply:
This approach is not working, as the `scheduler.shutdownNow()` is returning a "list of tasks that never commenced execution". So that will never be empty, as we will always have the next one scheduled. I don't know how could we add a sanity check here, but here is the documentation for `shutdownNow()`.
```
    /**
     * Attempts to stop all actively executing tasks, halts the
     * processing of waiting tasks, and returns a list of the tasks
     * that were awaiting execution.
     *
     * <p>This method does not wait for actively executing tasks to
     * terminate.  Use {@link #awaitTermination awaitTermination} to
     * do that.
     *
     * <p>There are no guarantees beyond best-effort attempts to stop
     * processing actively executing tasks.  For example, typical
     * implementations will cancel via {@link Thread#interrupt}, so any
     * task that fails to respond to interrupts may never terminate.
     *
     * @return list of tasks that never commenced execution
     * @throws SecurityException if a security manager exists and
     *         shutting down this ExecutorService may manipulate
     *         threads that the caller is not permitted to modify
     *         because it does not hold {@link
     *         java.lang.RuntimePermission}{@code ("modifyThread")},
     *         or the security manager's {@code checkAccess} method
     *         denies access.
     */
    List<Runnable> shutdownNow();
 
```

Reply:
I see. I thought the return value only includes "actively executing tasks" but not **scheduled-but-not-running-yet** threads. On another read it does.

Reply:
hmm. just for myself learning~ what does it mean "that will never be empty, as we will always have the next one scheduled"? 
what is "one"~?

Reply:
Previously in the `startSyncDataFromOM` method there is a `scheduler.scheduleWithFixedDelay`, which is doing this based (copied from the documentation):
```
Creates and executes a periodic action that becomes enabled first
after the given initial delay, and subsequently with the
given delay between the termination of one execution and the
commencement of the next.  If any execution of the task
encounters an exception, subsequent executions are suppressed.
Otherwise, the task will only terminate via cancellation or
termination of the executor.
```
Because of this there will always be a `syncDataFromOM()` scheduled with the declared delay, so the `shutdownNow()` method will return that.

Reply:
aha!! Thanks Zita!